### PR TITLE
Per-layer ffn hidden dims

### LINKED ISSUE #1361: [v2 FEATURE]: Layer-dependent hidden size for the feedforward neural network
## Is your feature request related to a problem? Please describe.

Following the MLPDS spring 2026 discussion, there is a request to allow each layer of the readout FFN to have its own hidden size, configurable from the CLI. Currently `--ffn-hidden-dim` takes a single int that is applied uniformly to every hidden layer, so users cannot try non-rectangular profiles (funnel, hourglass, etc.) without dropping into the Python API.
 
## Use-cases / examples
 
- **Funnel / decreasing widths** (e.g. `[600, 300, 150]`) — common pattern for tapered readout heads.
- **Hourglass / bottleneck** (e.g. `[512, 64, 512]`) — representational bottleneck as a regularizer for small datasets.
- **Reproducing literature baselines** that use non-uniform readout MLPs.
- **Architecture exploration** comparing uniform vs tapered FFNs at matched parameter count.
## Desired solution / workflow
 
Extend `--ffn-hidden-dim` to accept `nargs="+"`, preserving the single-int form for backward compatibility:
 
```bash
# Legacy (unchanged): uniform stack of width 300
chemprop train ... --ffn-num-layers 3 --ffn-hidden-dim 300
 
# New: explicit per-layer profile
chemprop train ... --ffn-hidden-dim 600 300 150
```
 
Sketch:
 
```python
parser.add_argument("--ffn-hidden-dim", type=int, nargs="+", default=[DEFAULT_HIDDEN_DIM])
 
# in build_model:
hidden_dims = args.ffn_hidden_dim
if len(hidden_dims) == 1:
    hidden_dims = hidden_dims * args.ffn_num_layers          # legacy path
else:
    if args.ffn_num_layers not in (None, len(hidden_dims)):
        raise ArgumentError("--ffn-num-layers must match len(--ffn-hidden-dim)")
    args.ffn_num_layers = len(hidden_dims)
```
 
The FFN builder in `chemprop/nn/ffn.py` would accept `hidden_dims: Sequence[int]` and iterate when stacking `nn.Linear` layers.

### PR DESCRIPTION
**Pull request by JacksonBurns**
> [!NOTE]
> This PR was handled entirely by Qwen 3.6 27b running locally via `opencode` with @JacksonBurns' supervision

Implements #1361 

~~Must be merged after #1367 (or merged into there)~~ undid this requirement in 228f5ee

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by JacksonBurns** [comment]
TODO: remove the unrelated changes to this PR so it can merged into `main`

**Comment by JacksonBurns** [comment]
@KnathanM I have dropped the unrelated changes, this can now be reviewed!

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -585,9 +593,39 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
+        if n_layers is not None and n_layers != len(hidden_dims):
+            raise ArgumentError(
+                argument=None,
+                message=(
+                    f"--{n_layers_key.replace('_', '-')}={n_layers} but "
+                    f"--{hidden_dim_key.replace('_', '-')} has {len(hidden_dims)} values. "
+                    "When providing multiple hidden dimensions, --ffn-num-layers must match "
+                    "the count or be omitted."
+                ),
+            )
```
Comment (KnathanM) [inline chemprop/cli/train.py:629]: Can you add an integration test? I'm thinking add `--ffn-hidden-dim 300 600` to test_train_quick in test_cli_regression_mol.py. When I run `chemprop train -i tests/data/regression/mol/mol.csv --epochs 3 --ffn-hidden-dim 300 600` I run into this ArgumentError because the default value for --ffn-num-layers is 1, so omitting it isn't an option. 

I'm okay with either requiring the user to pass --ffn-num-layers to match whenever they want hidden layers with different dims. Or maybe better is to just not give this warning `if n_layers != 1`. 

It may also be worth commenting in this error that --ffn-num-layers means the number of hidden layers in v2 (different from v1).

Comment (JacksonBurns) [reply chemprop/cli/train.py:629]: Good notes, I went with the 'make it a requirement to manually match' route in [dcdff5c](https://github.com/chemprop/chemprop/pull/1368/commits/dcdff5c08987d03fad4e290729deb3e6216f50b1)

File: tests/unit/cli/test_process_ffn_hidden_dims.py [hunk 2]
```diff
@@ -73,7 +73,15 @@ def test_all_ffn_variants_processed(self, base_args):
...
+        base_args.ffn_num_layers = 2
         _process_ffn_hidden_dims(base_args, "ffn_hidden_dim", "ffn_num_layers")
         assert base_args.ffn_hidden_dim == [256, 128]
         assert base_args.ffn_num_layers == 2
+
+    def test_mismatch_none_raises_error(self, base_args):
+        from configargparse import ArgumentError
+
+        base_args.ffn_hidden_dim = [600, 300, 150]
+        base_args.ffn_num_layers = None
```
Comment (KnathanM) [inline tests/unit/cli/test_process_ffn_hidden_dims.py]:
```suggestion
        base_args.ffn_num_layers = 1
```
the cli would never encounter None here right?

**Review by KnathanM — APPROVED** [review]
Good to merge. Could consider one small change before merging at your discretion.

File: tests/unit/cli/test_process_ffn_hidden_dims.py (hunk 2 above)
Comment (JacksonBurns) [reply tests/unit/cli/test_process_ffn_hidden_dims.py]: I think so, will commit and set to automerge!
